File: model/src/main/java/io/sparkled/model/animation/easing/reference/EasingTypes.java
package io.sparkled.model.animation.easing.reference;

import io.sparkled.model.animation.easing.EasingType;
import io.sparkled.model.animation.easing.EasingTypeCode;
import io.sparkled.model.animation.param.Param;
import io.sparkled.model.animation.param.ParamName;
import io.sparkled.model.animation.param.ParamType;

import java.util.Arrays;
import java.util.List;

public class EasingTypes {
    private static final List<EasingType> TYPES = Arrays.asList(
            easingType(EasingTypeCode.CONSTANT, "Constant",
                    param(ParamName.PERCENT, ParamType.DECIMAL).setValue(50f)
            ),
            easingType(EasingTypeCode.LINEAR, "Linear")
    );

    public static List<EasingType> get() {
        return TYPES;
    }

    private static EasingType easingType(EasingTypeCode easingType, String name, Param... params) {
        return new EasingType().setCode(easingType).setName(name).setParams(Arrays.asList(params));
    }

    private static Param param(ParamName paramName, ParamType type) {
        return new Param().setName(paramName.getName()).setType(type);
    }
}

File: model/src/main/kotlin/io/sparkled/model/animation/easing/reference/EasingTypes.kt
package io.sparkled.model.animation.easing.reference

import io.sparkled.model.animation.easing.EasingType
import io.sparkled.model.animation.easing.EasingTypeCode
import io.sparkled.model.animation.param.Param

object EasingTypes {
    private val TYPES = listOf(easingType(EasingTypeCode.EXPO_OUT), easingType(EasingTypeCode.LINEAR))

    fun get(): List<EasingType> {
        return TYPES
    }

    private fun easingType(code: EasingTypeCode, vararg params: Param): EasingType {
        return EasingType(code, code.displayName, listOf(*params))
    }
}
